### PR TITLE
[FIX] payment, website_sale, website_sale_picking: manage on site pick

### PR DESCRIPTION
**Pull request by thle-odoo**
Steps to reproduce:
-------------------
- install website_sale;
- install website_sale_picking;
- enable just "Pay in store when picking the product" provider;
- pusblish just "[On Site Pick] My Shop 1" shipping method;
- go to ecommerce and make the purchase flow.

Issue:
------
A traceback appears on the "/shop/payment" page.

Solution:
---------
The sale picking part must not manage payments.

opw-3451097

### CODE REVIEW

**Comment by AntoineVDV** [comment]
The codeowners will be requested again when the PR is ready for review.

File: addons/payment/static/src/js/checkout_form.js [hunk 1]
```diff
@@ -19,6 +19,7 @@
          * @constructor
          */
         init: function () {
+            this.txContext = {}; // It is important to initialize it before `start`
```
Comment (hubvd) [inline addons/payment/static/src/js/checkout_form.js:22]: This should also be done in  `manage_form.js` because it also uses the mixin. But I would do this
```diff
diff --git a/addons/payment/static/src/js/checkout_form.js b/addons/payment/static/src/js/checkout_form.js
index 51b17b4777be..6d8d194a7273 100644
--- a/addons/payment/static/src/js/checkout_form.js
+++ b/addons/payment/static/src/js/checkout_form.js
@@ -19,6 +19,7 @@
          * @constructor
          */
         init: function () {
+            paymentFormMixin.init.call(this);
             this.formType = 'checkout';
             const preventDoubleClick = handlerMethod => {
                 return debounce(handlerMethod, 500, true);
diff --git a/addons/payment/static/src/js/manage_form.js b/addons/payment/static/src/js/manage_form.js
index 9544bf30da1d..76d810c00ee3 100644
--- a/addons/payment/static/src/js/manage_form.js
+++ b/addons/payment/static/src/js/manage_form.js
@@ -24,6 +24,7 @@ import { debounce } from "@web/core/utils/timing";
          * @constructor
          */
         init: function () {
+            paymentFormMixin.init.call(this);
             this.formType = 'manage';
             const preventDoubleClick = handlerMethod => {
                 return debounce(handlerMethod, 500, true);
diff --git a/addons/payment/static/src/js/payment_form_mixin.js b/addons/payment/static/src/js/payment_form_mixin.js
index fae686c1aa93..15f113eab7cc 100644
--- a/addons/payment/static/src/js/payment_form_mixin.js
+++ b/addons/payment/static/src/js/payment_form_mixin.js
@@ -8,6 +8,10 @@
 
     export default {
 
+        init: function () {
+            this.txContext = {};
+        },
+
         /**
          * @override
          */

```

File: addons/payment/static/src/js/payment_form_mixin.js [hunk 2]
```diff
@@ -8,11 +8,14 @@
...
 
+        init: function () {
+            this.txContext = {};
+        },
+
         /**
          * @override
          */
         start: async function () {
-            this.txContext = {}; // Synchronously initialize txContext before any await.
```
Comment (vchu-odoo) [inline addons/payment/static/src/js/payment_form_mixin.js:15]: I believe that this line should not be changed except for the comment.
```py
this.txContext = {};
```

**Review by vchu-odoo — CHANGES_REQUESTED** [review]
The problem with this bug is not in payment but here https://github.com/odoo/odoo/blob/f185a9d6d1a004bf2ea111ef37f58155e7f0b173/addons/website_sale_picking/static/src/js/checkout_form.js#L79 It is not needed to click the payment option as it is handled in payment (if only one payment method is present, it is automatically selected). I suggest removing lines 78-79 from checkout_form and instead to correctly update the carrier_id of the preselected carrier in the start function of website_sale_delivery.js.

**Comment by pedrambiria** [comment]
@vchu-odoo Deleting line 79 from the checkout_form may resolve the current issue. However, it's important to consider the potential for future changes in payment behavior, which could introduce new issues here. To safeguard against this, it is needed to add a test in website_sale_picking to ensure the expected behavior remains consistent.

**Comment by AntoineVDV** [comment]
Hello, as said in https://github.com/odoo/odoo/pull/132834#pullrequestreview-1593090020, I would rather not have `payment` accommodate the bad code design choices made in other modules; a widget should neither make assumptions on the state of another widget nor try to control it. Instead, they should remain independent pieces.

It's true that forcing the initialization of the payment widget, adding tests, etc. would help safeguard against the class of bugs caused by one widget trying to take over another, but this widget interaction is actually the root cause of those bugs. For each similarly badly designed widget, there will exist another class of bugs to fix.

The best way to prevent those bugs is thus to stick to better practices and have widgets care about their own code and template.

tl;dr: fix with @vchu-odoo's suggestion. You can even remove the test because it won't prevent similar bad designs from breaking this flow again.

**Comment by AntoineVDV** [comment]
Hello, this bug will be fixed at the root with https://github.com/odoo/odoo/pull/138935 and https://github.com/odoo/enterprise/pull/49062 :)
Ideally, users should update the `payment` module for a better UX, but the fix works without that.
